### PR TITLE
Windows: Fix async restart data race/corruption

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

Fixes https://github.com/docker/docker/issues/26801.

### CODE REVIEW

**Comment by tonistiigi** [comment]
@jhowardmsft I think it was a mistake to put restartmanager to libcontainerd from daemon. If I understand the problem, it appears in linux as well and there is too much workarounds to make it work. Let me know it you agree, I can refactor the linux code to move it back.

**Comment by jhowardmsft** [comment]
@tonistiigi Sure, but regardless as it currently stands, we (I) have seen daemon lockups on Windows which are consistent with the problem as it currently stands. I've no problem doing some refactoring, but I would rather take this as a band-aid proven fix until that is available.

**Comment by tonistiigi** [comment]
@jhowardmsft Can you explain how this causes lockups?

**Comment by jhowardmsft** [comment]
libcontainerd ends up calling back into the daemon when the container lock isn't held. This leads to corruption of container state.

**Comment by jhowardmsft** [comment]
To add, I don't think Linux suffers from the same issue, but the code is quite divergent at this point due to the significant architectural differences.

File: daemon/monitor.go [hunk 1]
```diff
@@ -86,6 +86,20 @@ func (daemon *Daemon) StateChanged(id string, e libcontainerd.StateInfo) error {
 		} else {
 			logrus.Warnf("Ignoring StateExitProcess for %v but no exec command found", e)
 		}
+	case libcontainerd.StateAsyncStart:
+		// Start instigated asynchronously, such as in the case of restart on Windows.
+		// The container is not locked in this case
+		c.Lock()
+		defer c.Unlock()
+		c.SetRunning(int(e.Pid), e.State == libcontainerd.StateStart)
```
Comment (tonistiigi) [inline daemon/monitor.go]: Doesn't look correct as the param is always false.

**Comment by tonistiigi** [comment]
It's bad to add an extra state and a public create option for a completely internal behavior. The option is just used to pass a boolean from one method to another inside the library, it doesn't need to leak outside, nor does the restartmanager need to call public libcontainerd methods(it doesn't for linux).

I don't think this can cause lockups in linux(it can cause data races). If windows differs then we can probably find a fix that doesn't leak to portable areas. We should avoid different locking rules based on platform. https://github.com/docker/docker/pull/27149#issuecomment-251541410 is one case where a callback is made back to libcontainerd while holding a lock.

Still think correct solution is to move restartmanager.

**Comment by jhowardmsft** [comment]
Sure, but that is a much bigger fix. If you want to take that on....

File: daemon/monitor.go (hunk 1 above)
Comment (jhowardmsft) [inline daemon/monitor.go]: Yes, true. Copy/paste without reading the code from the StateStart case, not thinking that that also covered StateRestore. Updated.

**Comment by tonistiigi** [comment]
superseded by #27169

**Pull request closed without merging**